ros/src/waypoint_updater/nodes/updater.py

```diff
@@ -0,0 +1,227 @@
+from dynamic_reconfigure.server import Server
+from waypoint_updater.cfg import DynReconfConfig
+
+
+'''
+This node will publish waypoints from the car's current position to some `x`
+distance ahead.  As mentioned in the doc, you should ideally first implement
+a version which does not care about traffic lights or obstacles.
+
+Once you have created dbw_node, you will update this node to use the status of
+traffic lights too.
+
+Please note that our simulator also provides the exact location of traffic
+lights and their current status in `/vehicle/traffic_lights` message. You
+can use this message to build this node as well as to verify your TL
+classifier.
+
+'''
+
+
+class WaypointUpdater(object):
+    def __init__(self):
+        self.dyn_vals_received = False
+        self.waypoints = []
+        self.pose = None
+        self.velocity = None
+        self.final_waypoints = []
+        self.next_traffic_light_wp = None
+        if not self.dyn_vals_received:
+            self.update_rate = 3
+            self.default_velocity = 0.0
+            self.lookahead_wps = 50
+        self.subs = {}
+        self.pubs = {}
+
+        rospy.init_node('waypoint_updater')
+
+        self.subs['/current_pose'] = \
+            rospy.Subscriber('/current_pose', PoseStamped, self.pose_cb)
+
+        self.subs['/base_waypoints'] = \
+            rospy.Subscriber('/base_waypoints', Lane, self.waypoints_cb)
+
+        self.subs['/current_velocity'] = \
+            rospy.Subscriber('/current_velocity', TwistStamped,
+                             self.velocity_cb)
+
+        # TODO: Add a subscriber for /traffic_waypoint and /obstacle_waypoint
+        #  self.subs['/traffic_waypoint'] = \
+        #      rospy.Subscriber('/traffic_waypoint', String, self.traffic_cb)
+
+        self.pubs['/final_waypoints'] = rospy.Publisher('/final_waypoints',
+                                                        Lane, queue_size=1)
+
+        # TODO: Add other member variables you need below
+        self.dyn_reconf_srv = Server(DynReconfConfig, self.dyn_vars_cb)
+
+        self.loop()
+
+    def loop(self):
+        self.rate = rospy.Rate(self.update_rate)
+        # wait for waypoints and pose to be loaded before trying to update
+        # waypoints
+        while not self.waypoints:
+            self.rate.sleep()
+        while not self.pose:
+            self.rate.sleep()
+        while not rospy.is_shutdown():
+
+            if self.waypoints:
+                self.send_waypoints()
+            self.rate.sleep()
+
+    # adjust dynamic variables
+    def dyn_vars_cb(self, config, level):
+        if self.update_rate:
+            old_update_rate = self.update_rate
+            old_default_velocity = self.default_velocity
+            old_lookahead_wps = self.lookahead_wps
+        else:
+            self.dyn_vals_received = True
+            old_update_rate = 1
+            old_default_velocity = 0.0
+            old_lookahead_wps = 50
```

Comment:
I think this else condition for old values is not needed because they should always have been initialized with the default values.

Reply:
removed.